docs/proposals/parallel-compaction.md

```diff
@@ -0,0 +1,68 @@
+
+Once a compactor scheduler pushes a job to a compactor, the job is no longer available. Every set interval, or once the compaction is done, a compactor will update the compactor schedule the current status of the compaction job. If a compactor does not provide an update to the scheduler within a timeout, the compaction job becomes available to be assigned to other compactors.
+
+#### Concurrency
+
+To achieve concurrency within a single tenant, compactor scheduler will push jobs to compactors. Compactors are shuffle-sharded by tenant id, to prevent a large tenant from impacting the compaction of other tenants. Compactor will download blocks from long term storage, compact, and upload. Compactor will also pull from the clean up queues from scheduler, and delete blocks marked for deletion.
+
+#### Consistency
+
+On resharding of compactor schedulers, a tenant might move to a different scheduler. We can either drop the current compactor job in order to prevent duplicate compaction jobs, or continue compaction. I propose that the compactor drops the compaction job if the compaction group no longer belongs to the original compactor scheduler. This way, we do not have duplicate compactions happening, and we can minimize work wasted.
+
+### Contribute to Thanos for a more scalable compactor
+Instead of introducing parallelism on the Cortex compactor level, we move the parallelism to the Thanos compactor itself. Thanos has a [proposal to make compactor more scalable](https://docs.google.com/document/d/1xi0V8DB0hE54XgkogJRnNL6yH7C5JThJywlLFoC6dCQ/), and a [PR](https://github.com/thanos-io/thanos/pull/3807). Cortex will enjoy higher throughput per tenant if Thanos is able to speed up the compaction, and we can keep the Cortex architecture the same. However, this approach means that a single tenant is still sharded to a single compactor. In order to compact more groups at once, we must scale up compactor vertically. Although vertical scaling can get us far, we should scale horizontally where we can.
+
+### Create a Kubernetes job for each of the compaction plans
+The compactor schedules creates kubernetes jobs for each of the compaction jobs that can be done, instead of pushing the job to compactor. Kubernetes is responsible for spinning up a compactor pod for each of the job scheduled. This way, we avoid the coordination between compaction scheduler and compactor. Also, the compaction scheduler coordinates with the Kubernetes API server, instead of amongst each other during resharding. The downside of this approach is that we cannot control the concurrency, and Cortex requires a tighter integration with Kubernetes.
```

Comment:
> and Cortex requires a tighter integration with Kubernetes

Yeah, that's something we don't want. We're aware of users running Cortex outside K8S.